### ISSUE TITLE
[RFC] vectorize reducer  

### ISSUE THREAD

**Issue by chenqin** [dmlc/rabit#101]
Hi Folks,

We got some first hand results from running reducer in parallel with openmp. In light of hacky nature and thread "abuse" in previously proposed solution. I think it might worth a bit discussion before we move forward.

**Why we address this problem, really?**

Short answer is we have seen user running larger reducer exceed general guided depth 6-8. Per iteration speed got hit hard in distributed training as we sync on every depth on each iteration.

**What is the problem, precisely?**
![Screenshot from 2019-09-02 10-57-35](https://user-images.githubusercontent.com/1184534/64129392-8ac97300-cd70-11e9-8af7-23ff5457722e.png)
![Screenshot from 2019-09-02 10-57-07](https://user-images.githubusercontent.com/1184534/64129396-8c933680-cd70-11e9-883f-ec77a3fee1c0.png)
compiler will have hard time figure out data-flow and alignment when it tried to vectorize it.

**How do we plan to address this problem?**

We have 3 options

1. OpenMP simd pragma 
pro is this is a proven solution which might less likely see bugs or issues given it's well supported.
more detail here http://hpac.rwth-aachen.de/teaching/pp-16/material/08.OpenMP-4.pdf con is this would introduce openmp4 dependency.

2. use vector class https://github.com/vectorclass/version2 pro is we can use this as testing ground and optimize xgb as a whole. The con is this is a pretty new project, given it's experimental nature, we might run into various of bugs and issues.

3. hand tune loop and use compiler vectorization, pro is we control how we do it under gcc, con is this will make code less maintainable on various of platforms.
https://stackoverflow.com/questions/32000917/c-loop-optimization-help-for-final-assignment-with-compiler-optimization-disabl

For now, I would propose we choose least resistance path and go with openmp simd.

Thanks,
Chen

**Comment by hcho3** [comment]
@chenqin Is this related to `#pragma omp parallel` clause you inserted in #98 (and later removed)?

**Comment by chenqin** [comment]
> @chenqin Is this related to `#pragma omp parallel` clause you inserted in #98 (and later removed)?

that's right, threading seems hacky and error prone as you pointed out.

**Comment by trivialfis** [comment]
Can we try simple async first?

**Comment by chenqin** [comment]
Can we elaborate asynchronous suggestion more? We might look things differently.

Alllreduce semantic is synchronized blocking call. It’s generally higher performance if we allow higher out of order executions or simply drop slowest updates like PS does. 

However, this put ourselves in the risk of miss out key updates and affects one tree. Yes, those trees compliment each other so eventually might be okay. But we need to call out loud if we decide to do so.

For networking layer, we can definitely do reduce in streaming fashion. If that’s what you suggested.

**Comment by trivialfis** [comment]
I'm not sure which loop you want to vectorize.  Could you be more specific?  We can use async to achieve something similar to omp and in my experience in some cases it's faster than omp because we have explicit control over blocking.

**Comment by chenqin** [comment]
Let me try to explain how rabit network layer works. 

Every node has two set of connections tree toplogy and ring topology at same time.  When all parties do allreduce, depending on which topology user choose, every node receive and send data to it's connected peers based on ranking number. 

Of each given link, it has a buffer of 256MB by default, we use socket polling and check if a socket has data ingress or egress, if so, it will write to link buffer.
depending on the link with least data, it run reduce only on those chunk of data and back to listening mode until all data reduced. 

So it's event driven with socket polling and not blocked by slowest peer when traffic congestion happens. 

I think we might add async and reduce on out of sequence. It's not super clear how much it can help given most of polling already happens in kernel and low level where userspace only read flag from time to time and copy payload to buffer and run reduction.

The loop I want to vectorize does 152,524,800 x 2 simple customized reductions per node.
```
[94] allreduce (xgboost_build/src/tree/updater_histmaker.cc::739::CreateHist#16x152524800) finished version 12, seq 21, take 31.749563 seconds
```
It's a bit tricky to run full test without large scale job, some findings below

- a small performance increase by enable this on my local.(ubuntu amd ryzen 3)
- run gcc -fopt-info-loop-optimized=loop.miss and compare if loop vectorized

**Comment by chenqin** [comment]
https://github.com/dmlc/rabit/pull/102

**Comment by trivialfis** [comment]
@chenqin Thanks for your detailed explanation.  Looking through your PR helps.  Will take a closer look later.

**Comment by chenqin** [comment]
Updates:
we might also consider optimize serialize reduce as well. but it will require code change in xgboost side in order to boost speed. We can do another pr after this get in.

**Comment by CodingCat** [comment]
the code leading to the `We got some first hand results from running reducer in parallel with openmp. ` has been removed due to the  correctness issue

what's the current data-support as the motivation?

**Comment by chenqin** [comment]
> the code leading to the `We got some first hand results from running reducer in parallel with openmp. ` has been removed due to the correctness issue
> 
> what's the current data-support as the motivation?

Good point, working on getting datapoints. It's not urgent compare to boottrap_cache one

**Comment by hcho3** [comment]
Closing, as Rabit have been moved into [dmlc/xgboost](https://github.com/dmlc/xgboost). See discussion in [dmlc/xgboost#5995](https://github.com/dmlc/xgboost/issues/5995).
